src/v/ssx/watchdog.h

```diff
@@ -67,6 +67,39 @@ class watchdog {
       seastar::noncopyable_function<void()> callback) {
         start_waiting(timeout, std::move(callback));
     }
+
+    /// Create watchdog instance
+    ///
+    /// \param timeout defines time interval after which the watchdog will be
+    ///                triggered for the first time
+    /// \param repeat_interval defines the time interval after which the
+    ///                        watchdog will be trigger subsequently
+    /// \param callback is a callback that will be called when
+    ///                 the watchdog is triggered
+    /// \note The callback may outlive the watchdog instance. See the full note
+    /// on the previous constructor.
+    watchdog(
+      seastar::lowres_clock::duration timeout,
+      seastar::lowres_clock::duration repeat_interval,
+      seastar::noncopyable_function<void()> callback) {
+        ssx::background = ssx::ignore_shutdown_exceptions(seastar::do_with(
+          std::move(callback),
+          [this, timeout, repeat_interval](
+            seastar::noncopyable_function<void()>& cb) {
+              return seastar::sleep_abortable(timeout, _as)
+                .then([&cb] { cb(); })
+                .then([this, repeat_interval, &cb] {
+                    return seastar::repeat([this, repeat_interval, &cb] {
+                        return seastar::sleep_abortable(repeat_interval, _as)
```

Comment:
- if you call `cb()` as the first thing in the repeat block then return sleep abortable, can you get rid of an entire level of nesting?

- it's kinda nuanced, but the captured instance of `cb` is reinvoked. it's not correct or incorrect, but it might making sure that is aligned with other similar usages in repeats where the alternative is to capture a const-lambda such that cb() cannot modify itself.